CMakeLists.txt

```diff
@@ -163,6 +163,48 @@ option(BUILD_TESTING "Whether to enable tests" ON)
 
 option(BUILD_W3CTRACECONTEXT_TEST "Whether to build w3c trace context" OFF)
 
+option(MAINTAINER_MODE "Build in maintainer mode (-Wall -Werror)" OFF)
```

Comment:
> Just curious, what do we have in optional/separate maintainer mode? Is it possible to enforce `-Wall -Werror` for all cmake builds (maybe once we have fixed all warnings) ? 

Just a suggestion, could we add a option   to control whether to enable `-Werror` and `/WX` if we enable all these warnings for all cmake builds? When we upgrade a compiler, it usually brings more warnings, we don't want to break the compiling in package managers before upgrade the libraries, especially some warnings come from the dependency libraries.

Reply:
From experience, trying to enforce a warning free build in any combination of platforms / compilers / dependencies is bound to fail, because some warnings can be raised due to items not in the project control, such as:

- a system header file that causes a warning,
- a bug or limitation in a compiler, that causes a spurious warning,
- a dependency on a third party that causes a warning.

Instead of enforcing that any combination of platform / compiler / third party is warning free, a better approach is to:

- pick a given, recent and up to date, presumably clean, platform (here a recent linux for gcc and clang),
- pick a given, recent and up to date, presumably clean, compiler (default gcc and clang on a recent linux build),
- pick well known dependencies, recent if possible (here we take what is default for nlohmann_json, grpc, etc).

and enforce that the build is warning free on **_this_** combination.

Bugs reported and fixed will be beneficial for all.

Enforcing that any old/new platform with any old/new compiler and any old/new dependencies is warning free is counter productive, as it can not be achieved in practice, especially when the warning root cause is outside of the project source code.

Reply:
> > Just curious, what do we have in optional/separate maintainer mode? Is it possible to enforce `-Wall -Werror` for all cmake builds (maybe once we have fixed all warnings) ?
> 
> Just a suggestion, could we add a option to control whether to enable `-Werror` and `/WX` if we enable all these warnings for all cmake builds? When we upgrade a compiler, it usually brings more warnings, we don't want to break the compiling in package managers before upgrade the libraries, especially some warnings come from the dependency libraries.

@owent , I don't understand the suggestion itself.

There is a new flag to control -Werror and /WX, it is MAINTAINER_MODE=ON in CMake.

If we want to use it in every CMake build in the github repo, we can, by adding -DMAINTAINER_MODE=ON in the build script.

For package managers, there is no change at all: package managers are not expected to use MAINTAINER_MODE=ON, and the build without extra warning enforcement will work just as before.